File: Day77_78_79_SortAscendingDescending/ValidPalindrome/ValidPalindrome.cpp
class Solution {
public:
    bool isPalindrome(string s) {
        for(int i=0, j=s.size() - 1; i<j; i++, j--)  // Move two pointers from each end until they collide
        {
            while(isalnum(s[i]) == false && i<j) i++;  // If s[i] is not alphanumeric do i++
            while(isalnum(s[j]) == false && i<j) j--;  // If s[j] is not alphanumeric do j--
            if(toupper(s[i]) != toupper(s[j])) return false;
        }
        return true;
    }
};

// Time Complexity - O(n)
// Space Complexity - O(1)

File: Day77_78_79_SortAscendingDescending/ValidPalindrome/ValidPalindrome.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        l, r = 0, len(s) - 1
        while(l < r):
            while(l < r and s[l].isalnum() == False):
                l+=1
            while(l < r and s[r].isalnum() == False):
                r-=1
            if(s[l].lower() != s[r].lower()):
                return False
            l += 1
            r -= 1
        return True
        
# Time Complexity - O(n)
# Space Complexity - O(1)
